File: OsmAnd/src/net/osmand/plus/api/SQLiteAPIImpl.java
package net.osmand.plus.api;

import android.annotation.SuppressLint;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import androidx.annotation.Nullable;

import net.osmand.PlatformUtil;
import net.osmand.plus.OsmandApplication;

import org.apache.commons.logging.Log;

public class SQLiteAPIImpl implements SQLiteAPI {

	private final OsmandApplication app;
	private static final Log LOG = PlatformUtil.getLog(SQLiteAPIImpl.class);

	public SQLiteAPIImpl(OsmandApplication app) {
		this.app = app;
	}

	@SuppressLint("InlinedApi")
	@Nullable
	@Override
	public SQLiteConnection getOrCreateDatabase(String name, boolean readOnly) {
		android.database.sqlite.SQLiteDatabase db = null;
		try {
			db = app.openOrCreateDatabase(name, Context.MODE_PRIVATE
					| (readOnly ? 0 : Context.MODE_ENABLE_WRITE_AHEAD_LOGGING), null);
		} catch (RuntimeException e) {
			LOG.error(e.getMessage(), e);
		}
		if (db == null) {
			return null;
		}
		return new SQLiteDatabaseWrapper(db);
	}

	public class SQLiteDatabaseWrapper implements SQLiteConnection {
		android.database.sqlite.SQLiteDatabase ds;

		
		public SQLiteDatabaseWrapper(android.database.sqlite.SQLiteDatabase ds) {
			this.ds = ds;
		}

		@Override
		public int getVersion() {
			return ds.getVersion();
		}

		@Override
		public void close() {
			ds.close();
			
		}

		@Override
		public SQLiteCursor rawQuery(String sql, String[] selectionArgs) {
			Cursor c = ds.rawQuery(sql, selectionArgs);
			if(c == null) {
				return null;
			}
			return new SQLiteCursor() {
				
				@Override
				public boolean moveToNext() {
					return c.moveToNext();
				}
				
				@Override
				public String[] getColumnNames() {
					return c.getColumnNames();
				}
				
				@Override
				public boolean moveToFirst() {
					return c.moveToFirst();
				}
				
				@Override
				public String getString(int ind) {
					return c.getString(ind);
				}
				
				@Override
				public void close() {
					c.close();
				}
				
				public boolean isNull(int ind) {
					return c.isNull(ind);
				}

				@Override
				public int getColumnIndex(String columnName) {
					return c.getColumnIndex(columnName);
				}


				@Override
				public double getDouble(int ind) {
					return c.getDouble(ind);
				}
				

				@Override
				public long getLong(int ind) {
					return c.getLong(ind);
				}

				@Override
				public int getInt(int ind) {
					return c.getInt(ind);
				}

				@Override
				public byte[] getBlob(int ind) {
					return c.getBlob(ind);
				}
			};
		}

		@Override
		public void execSQL(String query) {
			ds.execSQL(query);			
		}

		@Override
		public void execSQL(String query, Object[] objects) {
			ds.execSQL(query, objects);
		}

		@Override
		public SQLiteStatement compileStatement(String query) {
			android.database.sqlite.SQLiteStatement st = ds.compileStatement(query);
			if(st == null) {
				return null;
			}
			return new SQLiteStatement() {
				
				@Override
				public void execute() {
					st.execute();
					
				}
				
				@Override
				public void close() {
					st.close();
				}
				
				@Override
				public void bindString(int i, String value) {
					st.bindString(i, value);
					
				}
				
				@Override
				public void bindNull(int i) {
					st.bindNull(i);
				}

				@Override
				public long simpleQueryForLong() {
					return st.simpleQueryForLong();
				}

				@Override
				public String simpleQueryForString() {
					return st.simpleQueryForString();
				}

				@Override
				public void bindLong(int i, long val) {
					st.bindLong(i, val);
				}

				@Override
				public void bindBlob(int i, byte[] val) {
					st.bindBlob(i, val);
				}
			};
		}

		@Override
		public void setVersion(int newVersion) {
			ds.setVersion(newVersion);
		}

		@Override
		public boolean isReadOnly() {
			return ds.isReadOnly();
		}

		@Override
		public boolean isClosed() {
			return !ds.isOpen();
		}
		
	}

	@Nullable
	@Override
	public SQLiteConnection openByAbsolutePath(String path, boolean readOnly) {
		// fix http://stackoverflow.com/questions/26937152/workaround-for-nexus-9-sqlite-file-write-operations-on-external-dirs
		android.database.sqlite.SQLiteDatabase db = SQLiteDatabase.openDatabase(path, null,
				readOnly ? SQLiteDatabase.OPEN_READONLY : (SQLiteDatabase.OPEN_READWRITE | SQLiteDatabase.ENABLE_WRITE_AHEAD_LOGGING));
		if (db == null) {
			return null;
		}
		return new SQLiteDatabaseWrapper(db);
	}
}

File: OsmAnd-shared/src/androidMain/kotlin/net/osmand/shared/api/SQLiteAPIImpl.kt
package net.osmand.shared.api

import android.content.Context
import android.content.Context.MODE_ENABLE_WRITE_AHEAD_LOGGING
import android.content.Context.MODE_PRIVATE
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteDatabase.CREATE_IF_NECESSARY
import android.database.sqlite.SQLiteDatabase.ENABLE_WRITE_AHEAD_LOGGING
import android.database.sqlite.SQLiteDatabase.OPEN_READONLY
import android.database.sqlite.SQLiteDatabase.OPEN_READWRITE
import net.osmand.shared.api.SQLiteAPI.SQLiteConnection
import net.osmand.shared.api.SQLiteAPI.SQLiteCursor
import net.osmand.shared.api.SQLiteAPI.SQLiteStatement
import net.osmand.shared.util.LoggerFactory
import java.io.File

class SQLiteAPIImpl(private val context: Context) : SQLiteAPI {

	companion object {
		private val log = LoggerFactory.getLogger("SQLiteAPIImpl")
	}

	override fun getOrCreateDatabase(name: String, readOnly: Boolean): SQLiteConnection? {
		val db = try {
			val file = context.getDatabasePath(name)
			if (readOnly && canOpenReadOnly(file)) {
				SQLiteDatabase.openDatabase(file.path, null, OPEN_READONLY)
			} else {
				// openOrCreateDatabase() always opens read-write. Keep write-ahead logging on for every
				// such open: a connection opened without it switches the file to a rollback journal,
				// the next one switches it back, each time under an exclusive lock that starves the
				// other connections until the busy timeout (OsmAnd-Issues #3331)
				context.openOrCreateDatabase(name, MODE_PRIVATE or MODE_ENABLE_WRITE_AHEAD_LOGGING, null)
			}
		} catch (e: RuntimeException) {
			log.error("Failed to get or create database", e)
			null
		}
		return db?.let { SQLiteDatabaseWrapper(it) }
	}

	/**
	 * A read-only connection can neither create the database nor roll back a hot journal left by
	 * an interrupted rollback-journal transaction; both need a writable open first.
	 */
	private fun canOpenReadOnly(file: File): Boolean =
		file.exists() && File(file.path + "-journal").length() == 0L

	inner class SQLiteDatabaseWrapper(private val ds: SQLiteDatabase) : SQLiteConnection {

		override fun getVersion(): Int = ds.version

		override fun close() = ds.close()

		override fun rawQuery(sql: String, selectionArgs: Array<String>?): SQLiteCursor? {
			val cursor = ds.rawQuery(sql, selectionArgs)
			return cursor.let { cursor ->
				object : SQLiteCursor {
					override fun moveToNext(): Boolean = cursor.moveToNext()
					override fun getColumnNames(): Array<String> = cursor.columnNames
					override fun moveToFirst(): Boolean = cursor.moveToFirst()
					override fun getString(ind: Int): String = cursor.getString(ind)
					override fun close() = cursor.close()
					override fun isNull(ind: Int): Boolean = cursor.isNull(ind)
					override fun getColumnIndex(columnName: String): Int = cursor.getColumnIndex(columnName)
					override fun getDouble(ind: Int): Double = cursor.getDouble(ind)
					override fun getLong(ind: Int): Long = cursor.getLong(ind)
					override fun getInt(ind: Int): Int = cursor.getInt(ind)
					override fun getBlob(ind: Int): ByteArray = cursor.getBlob(ind)
				}
			}
		}

		override fun execSQL(query: String) = ds.execSQL(query)

		override fun execSQL(query: String, objects: Array<Any?>) = ds.execSQL(query, objects)

		override fun compileStatement(query: String): SQLiteStatement {
			val st = ds.compileStatement(query)
			return st.let { statement ->
				object : SQLiteStatement {
					override fun execute() = statement.execute()
					override fun close() = statement.close()
					override fun bindString(i: Int, value: String) = statement.bindString(i, value)
					override fun bindNull(i: Int) = statement.bindNull(i)
					override fun simpleQueryForLong(): Long = statement.simpleQueryForLong()
					override fun simpleQueryForString(): String = statement.simpleQueryForString()
					override fun bindLong(i: Int, value: Long) = statement.bindLong(i, value)
					override fun bindBlob(i: Int, value: ByteArray) = statement.bindBlob(i, value)
					override fun bindDouble(i: Int, value: Double) = statement.bindDouble(i, value)
				}
			}
		}

		override fun setVersion(newVersion: Int) {
			ds.version = newVersion
		}

		override fun isReadOnly(): Boolean = ds.isReadOnly

		override fun isClosed(): Boolean = !ds.isOpen

		override fun beginTransaction() {
			ds.beginTransaction()
		}

		override fun setTransactionSuccessful() {
			ds.setTransactionSuccessful()
		}

		override fun endTransaction() {
			ds.endTransaction()
		}
	}

	override fun openByAbsolutePath(path: String, readOnly: Boolean): SQLiteConnection? {
		return try {
			val flags = if (readOnly) OPEN_READONLY else (OPEN_READWRITE or ENABLE_WRITE_AHEAD_LOGGING or CREATE_IF_NECESSARY)
			return SQLiteDatabaseWrapper(SQLiteDatabase.openDatabase(path, null, flags))
		} catch (e: Exception) {
			log.error("Failed to open database by path: $path readOnly=$readOnly", e)
			null
		}
	}
}